### DISCUSSION TITLE
The saved flow file is garbled

### DISCUSSION THREAD

**Discussion by WaterLoran** [mitmproxy/mitmproxy#6275] [General]
#### Problem Description
After I grabbed some packages and downloaded them, after checking, the information in the files is garbled

#### Steps to reproduce the behavior:
1. Click on the web page
2. Click save
3. Open the downloaded flow file
4. The file is garbled, and after trying ANSI, UTF-8 encoding format, it is garbled

#### System Information
Mitmproxy: 9.0.1
Python:    3.9.7
OpenSSL:   OpenSSL 3.0.7 1 Nov 2022
Platform:  Windows-10-10.0.19045-SP0

**Comment by mhils** [comment]
Saved flow files are not meant to be human-readable. You can open them
again with mitmproxy. :)

**Comment by WaterLoran** [comment]
Hello! Thank you very much for your timely response in the other hemisphere
I am very happy and excited, after all, this is my first time communicating with an open source author

Firstly, I understand the feature you mentioned, and I also know that after importing it to the software platform, you can reopen it to view it

Secondly, I am a Test automation engineer
However, my actual request is to download all the requested information and then parse it,
Keyword for generating automation, and automatic generation of automation scripts (dependent on the context of all requests)

In fact, there is a function that is quite similar, but it still doesn't meet my needs. It can only download one requested data, and that function is "Flow=>Export=>Export raw request and response"

So, my actual expectation is that one of the following
1. Able to export all Flow information in text format, making it easy for me to parse
2. Able to export all information through the "export" function, rather than just one item

Note: In fact, I have also tried using mitmweb - s xxx.py to extract information, but there is a problem with this processing method, which may require additional processing of invalid Flow streams, which may take me some time

您好! 非常感谢您在另一个半球的及时回复.
我表示非常开心和激动, 毕竟这是我第一次和开源作者沟通.

首先,我理解你说的那个功能, 我也知道导入到软件平台之后可以重新打开查看.

其次我是一个自动化测试工程师
但是,我的实际诉求是下载所有请求的信息, 然后解析这些信息, 
用于生成自动化的关键字, 和自动生成自动化脚本(依赖于所有的请求的上下文)

实际上有一个功能比较类似, 但还不满足我 ,他只能下载一个请求的数据,那个功能就是"FLow=>Export=>Export raw request and response".

所以,我的实际期望是, 下面的其中一个
1. 能够导出所有Flow的信息, 并且是文本格式的, 这样子很方便我解析
2. 能够通过"export"功能导出所有的信息, 而非只有一条

备注: 实际上我也尝试过, 使用mitmerb -s xxx.py来提取信息, 但是这样子的处理方式会有一个问题, 那就是有可能需要对无效的Flow流进行额外的处理, 这可能会花费我一点时间.

**Comment by mhils** [comment]
Take a look at https://github.com/mitmproxy/mitmproxy/blob/main/examples/addons/io-read-saved-flows.py. This should cover your use case!

**Comment by WaterLoran** [comment]
我按照你的方法去执行了脚本, 但是实际上,并不能使用该及哦啊本将服务拉起来
I executed the script according to your method, but in reality, I cannot use it to pull up the service

command::  mitmweb -s .\io-read-saved-flows.py

Loading script .\io-read-saved-flows.py
in script .\io-read-saved-flows.py: [Errno 2] No such file or directory: '-s'
Web server listening at http://127.0.0.1:8081/                               
Proxy server listening at http://*:8080
Error on reading from the event loop self pipe                 
loop: <ProactorEventLoop running=True closed=False debug=False>
Traceback (most recent call last):                             
  File "c:\python\lib\asyncio\runners.py", line 43, in run     
    return loop.run_until_complete(main)                       
  File "c:\python\lib\asyncio\base_events.py", line 603, in run_until_complete
    self.run_forever()
  File "c:\python\lib\asyncio\windows_events.py", line 316, in run_forever
    super().run_forever()
  File "c:\python\lib\asyncio\base_events.py", line 570, in run_forever
    self._run_once()
  File "c:\python\lib\asyncio\base_events.py", line 1859, in _run_once
    handle._run()
  File "c:\python\lib\asyncio\events.py", line 81, in _run
SystemExit: 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:\python\lib\asyncio\proactor_events.py", line 768, in _loop_self_reading
    f.result()  # may raise
  File "c:\python\lib\asyncio\windows_events.py", line 808, in _poll
    value = callback(transferred, key, ov)
  File "c:\python\lib\asyncio\windows_events.py", line 457, in finish_recv
    raise ConnectionResetError(*exc.args)
ConnectionResetError: [WinError 995] 由于线程退出或应用程序请求，已中止 I/O 操作。
Task exception was never retrieved
future: <Task finished name='Task-7' coro=<ErrorCheck._shutdown_if_errored() done, defined at c:\python\lib\site-packages\mitmproxy\addons\errorcheck.py:18> exception=SystemExit(1)>
Traceback (most recent call last):
  File "c:\python\lib\asyncio\runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "c:\python\lib\asyncio\base_events.py", line 603, in run_until_complete
    self.run_forever()
  File "c:\python\lib\asyncio\windows_events.py", line 316, in run_forever
    super().run_forever()
  File "c:\python\lib\asyncio\base_events.py", line 570, in run_forever
    self._run_once()
  File "c:\python\lib\asyncio\base_events.py", line 1859, in _run_once
    handle._run()
  File "c:\python\lib\asyncio\events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
  File "c:\python\lib\site-packages\mitmproxy\addons\errorcheck.py", line 20, in _shutdown_if_errored
    sys.exit(1)
SystemExit: 1

**Comment by WaterLoran** [comment]
To think about all the requested information, I think I have found a way

That's it

1. Through http://127.0.0.1:8081/flows Obtain information about all flows and extract the IDs of all flows

2. Then, based on the IDs of these streams, go to the http://127.0.0.1:8081/commands/export This interface can obtain all request and response data by passing in the specified ID

**Comment by WaterLoran** [comment]
When I finish writing it, if necessary, I will also provide it as an example, hoping to adopt it, haha

**Comment by nneonneo** [comment]
`io-read-saved-flows.py` isn't an addon (even though it's in the addons directory); rather, it's a standalone script which can be run from Python directly (if mitmproxy is installed), i.e. `python io-read-saved-flows.py <flowfile>`.
